Add eq3btsmart integration

EuleMitKeule opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This PR reintroduces the `eq3btsmart` integration that allows control and communication with [eQ-3 Bluetooth Smart Thermostat](https://www.eq-3.de/produkte/eqiva/detail/bluetooth-smart-heizkoerperthermostat.html) devices.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/31202
- Link to brands pull request: https://github.com/home-assistant/brands/pull/5160

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

EuleMitKeule committed:
Add eq3btsmart platform

EuleMitKeule committed:
Reduce to single platform

EuleMitKeule committed:
Merge branch 'dev' into eq3btsmart

EuleMitKeule committed:
Add eq3btsmart to brands/eq3.json

EuleMitKeule committed:
Merge branch 'dev' into eq3btsmart

EuleMitKeule commented:
I have reduced the changes to a single platform.

zweckj commented:
I haven't looked at your PR in detail, but some suggestions:

- check your `strings.json` for things you don't use atm
- the config flow needs to have 100% test coverage
- move the options flow to a separate PR
- both `models.py` and ``schemas.py` look like they below in the pypi library
- this whole PR is **HUGE**, so think about things you can do to make it more reviewable (e.g. moving calculations in the `climate.py` to the lib)

EuleMitKeule commented:
@zweckj Thanks for taking a look!

> check your `strings.json` for things you don't use atm
Yeah, I'll remove the unused strings. They are for services that will be added once I submit a PR for the other platforms.

> the config flow needs to have 100% test coverage
I thought this was only a requirement to get a gold quality rating, but I guess I can add tests for that.

> move the options flow to a separate PR
Didn't think that was necessary, but I can do that as well.

> both `models.py` and `schemas.py` look like they below in the pypi library
Actually the models are just some simple dataclasses for the data provided in the config and options flows and have nothing to do with the library logic. The `schemas.py` contains the voluptuous schemas for the config and options flows.

> this whole PR is **HUGE**, so think about things you can do to make it more reviewable (e.g. moving calculations in the `climate.py` to the lib)
I don't think there is much I can change about that fact. Climate entities are just pretty complex. Maybe I can spare a few lines.

zweckj commented on homeassistant/components/eq3btsmart/climate.py:
```diff
@@ -0,0 +1,422 @@
+            case Preset.AWAY:
+                await self._thermostat.async_set_away(True)
+            case Preset.ECO:
+                if self._thermostat.status is None:
+                    return
+                if self._thermostat.status.is_boost:
+                    await self._thermostat.async_set_boost(False)
+                if self._thermostat.status.is_away:
+                    await self._thermostat.async_set_away(False)
```
this could be a function

EuleMitKeule replied:
Done, logic is now implemented in the library.

zweckj commented on homeassistant/components/eq3btsmart/const.py:
```diff
@@ -0,0 +1,113 @@
+CONF_EXTERNAL_TEMP_SENSOR = "conf_external_temp_sensor"
+CONF_RSSI = "rssi"
+
+ENTITY_NAME_BUSY = "Busy"
+ENTITY_NAME_CONNECTED = "Connected"
+ENTITY_NAME_BATTERY = "Battery"
+ENTITY_NAME_WINDOW_OPEN = "Window Open"
+ENTITY_NAME_DST = "dSt"
+ENTITY_NAME_FETCH_SCHEDULE = "Fetch Schedule"
+ENTITY_NAME_FETCH = "Fetch"
+ENTITY_NAME_LOCKED = "Locked"
+ENTITY_NAME_COMFORT = "Comfort"
+ENTITY_NAME_ECO = "Eco"
+ENTITY_NAME_OFFSET = "Offset"
+ENTITY_NAME_WINDOW_OPEN_TEMPERATURE = "Window Open"
+ENTITY_NAME_WINDOW_OPEN_TIMEOUT = "Window Open Timeout"
+ENTITY_NAME_AWAY_HOURS = "Away Hours"
+ENTITY_NAME_AWAY_TEMPERATURE = "Away"
+ENTITY_NAME_VALVE = "Valve"
+ENTITY_NAME_AWAY_END = "Away until"
+ENTITY_NAME_RSSI = "Rssi"
+ENTITY_NAME_SERIAL_NUMBER = "Serial"
+ENTITY_NAME_FIRMWARE_VERSION = "Firmware Version"
+ENTITY_NAME_MAC = "MAC"
+ENTITY_NAME_RETRIES = "Retries"
+ENTITY_NAME_PATH = "Path"
+ENTITY_NAME_AWAY_SWITCH = "Away"
+ENTITY_NAME_BOOST_SWITCH = "Boost"
+ENTITY_NAME_CONNECTION = "Connection"
+ENTITY_NAME_MONITORING = "Monitoring"
```
where do you use those? Those probably belong to `strings.json`

EuleMitKeule replied:
Those are the names for the entities that I had to remove for now since I am only allowed to add a single platform in this PR. I will remove them.

EuleMitKeule replied:
Removed for now since these are not used by the climate platform.

zweckj commented on homeassistant/components/eq3btsmart/const.py:
```diff
@@ -0,0 +1,113 @@
+ENTITY_NAME_CONNECTION = "Connection"
+ENTITY_NAME_MONITORING = "Monitoring"
+
+ENTITY_ICON_VALVE = "mdi:pipe-valve"
+ENTITY_ICON_AWAY_SWITCH = "mdi:lock"
+ENTITY_ICON_BOOST_SWITCH = "mdi:speedometer"
+ENTITY_ICON_CONNECTION = "mdi:bluetooth"
```
and those to `icons.json`

EuleMitKeule replied:
Removed for now since these are not used by the climate platform.

zweckj commented on homeassistant/components/eq3btsmart/climate.py:
```diff
@@ -0,0 +1,422 @@
+                    return None
+
+                return (
+                    (1 - self._thermostat.status.valve / 100) * 2
+                    + self._thermostat.status.target_temperature.value
+                    - 2
```
I think we're not supposed to do calculations in our entites, but only in the lib

EuleMitKeule replied:
Done, calculation is now implemented in the library.

zweckj commented:
> Actually the models are just some simple dataclasses for the data provided in the config and options flows and have nothing to do with the library logic. The `schemas.py` contains the voluptuous schemas for the config and options flows.

Oh you're right, I just took a quick glance...

EuleMitKeule committed:
Remove unused constants, remove calculations

EuleMitKeule committed:
Remove options flow, remove unused schemas and strings, bump library version

EuleMitKeule committed:
Add tests for config flow

EuleMitKeule commented:
@zweckj I have added tests for the config flow, removed the options flow and removed unused strings and constants. Would be nice if you could take another look!

EuleMitKeule committed:
Fix typing issues

EuleMitKeule committed:
Improve manifest

EuleMitKeule committed:
Remove unused options flow strings

zweckj commented on homeassistant/components/eq3btsmart/__init__.py:
```diff
@@ -0,0 +1,153 @@
+
+    mac_address: str = entry.data[CONF_MAC]
+    name: str = entry.data[CONF_NAME]
+    adapter: Adapter = entry.options.get(CONF_ADAPTER, DEFAULT_ADAPTER)
+    current_temp_selector = entry.options.get(
+        CONF_CURRENT_TEMP_SELECTOR, DEFAULT_CURRENT_TEMP_SELECTOR
+    )
+    target_temp_selector = entry.options.get(
+        CONF_TARGET_TEMP_SELECTOR, DEFAULT_TARGET_TEMP_SELECTOR
+    )
+    external_temp_sensor = entry.options.get(CONF_EXTERNAL_TEMP_SENSOR, "")
+    scan_interval = entry.options.get(CONF_SCAN_INTERVAL, DEFAULT_SCAN_INTERVAL)
```
those are currently empty then, aren't they?

EuleMitKeule replied:
Yep you're right, will remove.

zweckj commented on homeassistant/components/eq3btsmart/__init__.py:
```diff
@@ -0,0 +1,153 @@
+
+    eq3_config_entry = Eq3ConfigEntry(eq3_config=eq3_config, thermostat=thermostat)
+
+    domain_data: dict[str, Any] = hass.data.setdefault(DOMAIN, {})
+    domain_data[entry.entry_id] = eq3_config_entry
```
```suggestion
    hass.data.setdefault(DOMAIN, {})[entry.entry_id] =  eq3_config_entry
```

zweckj commented on homeassistant/components/eq3btsmart/climate.py:
```diff
@@ -0,0 +1,390 @@
+    """Handle config entry setup."""
+
+    eq3_config_entry: Eq3ConfigEntry = hass.data[DOMAIN][config_entry.entry_id]
+    thermostat = eq3_config_entry.thermostat
+    eq3_config = eq3_config_entry.eq3_config
+
+    entities_to_add: list[Entity] = [Eq3Climate(eq3_config, thermostat)]
+
+    async_add_entities(
+        entities_to_add,
+        update_before_add=False,
+    )
```
```suggestion
    async_add_entities(
        [Eq3Climate(eq3_config_entry.eq3_config, eq3_config_entry.thermostat,)]
    )
```

EuleMitKeule replied:
I normally like to be a bit more expressive and have this in separate lines, especially since in the platforms that will be added in the future some entities are only created based on a flag set in the options flow, but I guess I'll simplify this in the way you suggested.

zweckj commented on homeassistant/components/eq3btsmart/climate.py:
```diff
@@ -0,0 +1,390 @@
+        if temperature is None:
+            return
+
+        temperature = round(temperature * 2) / 2
+        temperature = min(temperature, self.max_temp)
+        temperature = max(temperature, self.min_temp)
```
what exactly are you doing this for?

EuleMitKeule replied:
The thermostats can only be set to temperatures that are increments of 0.5, so we round the temperature to the nearest half and also check that it is inside the temperature limits set by the thermostat. I guess this could also be extracted into the library.

zweckj commented on homeassistant/components/eq3btsmart/climate.py:
```diff
@@ -0,0 +1,390 @@
+        if self._thermostat.status is not None:
+            self._target_temperature = self._thermostat.status.target_temperature.value
+
+    @property
+    def device_info(self) -> DeviceInfo | None:
+        """Return device information."""
+
+        return DeviceInfo(
+            name=self._eq3_config.name,
+            manufacturer=MANUFACTURER,
+            model=DEVICE_MODEL,
+            identifiers={(DOMAIN, self._eq3_config.mac_address)},
+            sw_version=self._firmware_version,
+            connections={(CONNECTION_BLUETOOTH, self._eq3_config.mac_address)},
+        )
```
personal preference: set it thorugh `_attr_device_info` as it's static

EuleMitKeule replied:
True, I only found out about that attribute today, will use that.

zweckj commented on homeassistant/components/eq3btsmart/config_flow.py:
```diff
@@ -0,0 +1,94 @@
+class EQ3ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Config flow for eQ-3 Bluetooth Smart thermostats."""
+
+    VERSION = 1
```
```suggestion
```

is the default

zweckj commented on homeassistant/components/eq3btsmart/config_flow.py:
```diff
@@ -0,0 +1,94 @@
+    ) -> FlowResult:
+        """Handle a flow initialized by the user."""
+
+        errors: dict[str, str] | None = {}
```
```suggestion
```
not used, but is there a way to validate the user entered configuration is even valid?

EuleMitKeule committed:
Simplify code, validate user input

EuleMitKeule committed:
Fix device info, add bluetooth device

bdraco commented on homeassistant/components/eq3btsmart/manifest.json:
```diff
@@ -0,0 +1,27 @@
+      "connectable": true
+    }
+  ],
+  "codeowners": ["@eulemitkeule", "@dbuezas"],
```
It would great to get a 👍  from dbuezas here before merging this

EuleMitKeule replied:
Sure, I wrote him a DM and asked if he wanted to be listed before, told him to react here now.

dbuezas replied:
I'm ok with being a co-code owner👍
Awesome work @EuleMitKeule

bdraco commented on homeassistant/components/eq3btsmart/climate.py:
```diff
@@ -0,0 +1,391 @@
+            return Preset.LOW_BATTERY
+        if self._thermostat.status.is_away:
+            return Preset.AWAY
+        if self._thermostat.status.operation_mode == OperationMode.ON:
```
```suggestion
        if self._thermostat.status.operation_mode is OperationMode.ON:
```

bdraco replied:
If these are both enums than identity checks are preferred

bdraco commented on homeassistant/components/eq3btsmart/climate.py:
```diff
@@ -0,0 +1,391 @@
+        if self._thermostat.status is None:
+            return None
+
+        if self._thermostat.status.operation_mode == OperationMode.OFF:
```
```suggestion
        if self._thermostat.status.operation_mode is OperationMode.OFF:
```

bdraco replied:
If these are both enums than identity checks are preferred

EuleMitKeule committed:
Use identity checks for enums

EuleMitKeule commented:
The CI was not executed due to rate limitation it seems.

bdraco committed:
Merge branch 'dev' into eq3btsmart

bdraco commented:
Beta week so github is very busy. I merged in `dev`. It will try again. If its still rate limited, I'll start it again later

bdraco commented on homeassistant/components/eq3btsmart/__init__.py:
```diff
@@ -0,0 +1,124 @@
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Handle config entry unload."""
+
+    unload_ok = await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
+
+    if unload_ok:
```
```suggestion
    if unload_ok := await hass.config_entries.async_unload_platforms(entry, PLATFORMS):
```

bdraco commented on homeassistant/components/eq3btsmart/__init__.py:
```diff
@@ -0,0 +1,124 @@
+
+    device: BLEDevice | None
+
+    if config.adapter == Adapter.AUTO:
```
If this is an enum on both sides, please use identity checks

bdraco commented on homeassistant/components/eq3btsmart/climate.py:
```diff
@@ -0,0 +1,391 @@
+            _LOGGER.warning("[%s] Disconnected", self._eq3_config.name)
+            self._was_connected = False
+
+        self.schedule_update_ha_state()
```
Please make sure you are using the async api in `@callback` and `async` functions.

https://github.com/home-assistant/core/blob/52d27230bce239017722d8ce9dd6f5386f63aba2/homeassistant/helpers/entity.py#L1207

bdraco commented on homeassistant/components/eq3btsmart/climate.py:
```diff
@@ -0,0 +1,391 @@
+            )
+
+    @callback
+    def _on_connection_changed(self, is_connected: bool = True) -> None:
```
```suggestion
    def _async_on_connection_changed(self, is_connected: bool = True) -> None:
```

Please prefix functions that are async safe and decorated with `@callback` with `async` so we know they are intended to run in the event loop as we have both sync and async code in this code base

bdraco commented on homeassistant/components/eq3btsmart/climate.py:
```diff
@@ -0,0 +1,391 @@
+        if self._thermostat.device_data is not None:
+            self._on_device_updated()
+
+        self.schedule_update_ha_state()
```
Please make sure you are using the async api in `@callback` and `async` functions.

https://github.com/home-assistant/core/blob/52d27230bce239017722d8ce9dd6f5386f63aba2/homeassistant/helpers/entity.py#L1207

bdraco commented on homeassistant/components/eq3btsmart/climate.py:
```diff
@@ -0,0 +1,391 @@
+            )
+
+    @callback
+    def _on_updated(self) -> None:
```
```suggestion
    def _async_on_updated(self) -> None:
```

Please prefix functions that are async safe and decorated with `@callback` with `async` so we know they are intended to run in the event loop as we have both sync and async code in this code base

bdraco commented on homeassistant/components/eq3btsmart/climate.py:
```diff
@@ -0,0 +1,391 @@
+
+        await self.async_scan()
+
+        if self._platform_state != EntityPlatformState.REMOVED:
```
```suggestion
        if self._platform_state is not EntityPlatformState.REMOVED:
```

bdraco replied:
Its not indented to access `self._platform_state`

This should be cancelled when the entity is removed from hass instead.

Ideally this logic does not live in the entity

bdraco commented on homeassistant/components/eq3btsmart/climate.py:
```diff
@@ -0,0 +1,391 @@
+            )
+            return
+
+    @property
+    def available(self) -> bool:
+        """Return True if entity is available."""
+
+        return self._is_available
```
Looks like this would be simpler to implement as `self._attr_available`

bdraco commented on homeassistant/components/eq3btsmart/climate.py:
```diff
@@ -0,0 +1,391 @@
+        except ValueError as ex:
+            raise ServiceValidationError("Invalid temperature") from ex
+
+    @property
+    def hvac_mode(self) -> HVACMode | None:
+        """Return the current HVAC mode."""
+
+        return self._attr_hvac_mode
+
```
This looks like unneeded code

bdraco commented on homeassistant/components/eq3btsmart/climate.py:
```diff
@@ -0,0 +1,391 @@
+
+        return self._is_available
+
+    @property
+    def hvac_action(self) -> HVACAction | None:
```
It looks like most of these properties could be set as `_attr_*` on update instead

bdraco commented on homeassistant/components/eq3btsmart/climate.py:
```diff
@@ -0,0 +1,391 @@
+        if self._thermostat.status is not None:
+            self._target_temperature = self._thermostat.status.target_temperature.value
+
+    @property
+    def device_info(self) -> DeviceInfo | None:
+        """Return device information."""
+
+        return DeviceInfo(
+            name=self._eq3_config.name,
+            manufacturer=MANUFACTURER,
+            model=DEVICE_MODEL,
+            identifiers={(DOMAIN, self._eq3_config.mac_address)},
+            sw_version=self._firmware_version,
+            connections={(CONNECTION_BLUETOOTH, self._eq3_config.mac_address)},
+        )
```
```suggestion
```

already set https://github.com/home-assistant/core/pull/109291/files#diff-ba702aab1c1cc3e0db5a6529b2bd0d2777be0f1568318bb3b5a36205ecf1c137R98

bdraco commented on homeassistant/components/eq3btsmart/climate.py:
```diff
@@ -0,0 +1,391 @@
+            await self._thermostat.async_get_status()
+        except Eq3Exception:
+            self._is_available = False
+            self.schedule_update_ha_state()
```
Please see the async api in async code

bdraco commented on homeassistant/components/eq3btsmart/models.py:
```diff
@@ -0,0 +1,39 @@
+)
+
+
+@dataclass
```
```suggestion
@dataclass(slots=True)
```

bdraco commented on homeassistant/components/eq3btsmart/models.py:
```diff
@@ -0,0 +1,39 @@
+    default_away_temperature: float = DEFAULT_AWAY_TEMP
+
+
+@dataclass
```
```suggestion
@dataclass(slots=True)
```

bdraco commented on tests/components/eq3btsmart/test_config_flow.py:
```diff
@@ -0,0 +1,150 @@
+        )
+        await hass.async_block_till_done()
+
+    assert result["type"] == FlowResultType.FORM
+    assert result["errors"] == {CONF_NAME: "invalid_name"}
+    assert len(mock_setup_entry.mock_calls) == 0
+
```
Please finish all config flow tests to reach CREATE_ENTRY or ABORT states to ensure they can recover from failures

bdraco commented on tests/components/eq3btsmart/test_config_flow.py:
```diff
@@ -0,0 +1,150 @@
+        )
+        await hass.async_block_till_done()
+
+    assert result["type"] == FlowResultType.FORM
+    assert result["errors"] == {CONF_MAC: "invalid_mac_address"}
+    assert len(mock_setup_entry.mock_calls) == 0
```
Please finish all config flow tests to reach CREATE_ENTRY or ABORT states to ensure they can recover from failures

bdraco commented on homeassistant/components/eq3btsmart/climate.py:
```diff
@@ -0,0 +1,391 @@
+        self._is_available = False
+        self._cancel_timer: Callable[[], None] | None = None
+        self._attr_has_entity_name = True
+        self._attr_name = None
```
Why set the name to none?

EuleMitKeule replied:
This is done so that `use_device_name` will return `True` and the device name is used as the name for the climate entity.

bdraco replied:
Right since you are using `_attr_has_entity_name` it should be `None`

Please make them class attributes

bdraco replied:
https://github.com/home-assistant/core/pull/109291/files#r1480070134

bdraco commented on homeassistant/components/eq3btsmart/climate.py:
```diff
@@ -0,0 +1,391 @@
+        self._cancel_timer: Callable[[], None] | None = None
+        self._attr_has_entity_name = True
+        self._attr_name = None
+        self._attr_hvac_mode = None
```
```suggestion
```

This is the default

bdraco replied:
Right, its only `hvac_action` that is initialized with a default

https://github.com/home-assistant/core/blob/545b888034966f45127477bed5a9cf2fa08a9730/homeassistant/components/climate/__init__.py#L283

bdraco commented on homeassistant/components/eq3btsmart/climate.py:
```diff
@@ -0,0 +1,391 @@
+            name=self._eq3_config.name,
+            manufacturer=MANUFACTURER,
+            model=DEVICE_MODEL,
+            identifiers={(DOMAIN, self._eq3_config.mac_address)},
```
```suggestion
```

identifiers is not needed if connections are set

bdraco commented on homeassistant/components/eq3btsmart/climate.py:
```diff
@@ -0,0 +1,391 @@
+    async def async_added_to_hass(self) -> None:
+        """Run when entity about to be added to hass."""
+
+        asyncio.get_event_loop().create_task(self._async_scan_loop())
```
Please use the Home Assistant api to create the task

`hass.async_create_bakcground_task` or `hass.async_create_task`

Creating a task without holding a strong reference to it may result in it being garbage collected before it finishes

https://docs.python.org/3/library/asyncio-task.html#asyncio.create_task

> Important Save a reference to the result of this function, to avoid a task disappearing mid-execution. The event loop only keeps weak references to tasks. A task that isn’t referenced elsewhere may get garbage collected at any time, even before it’s done. For reliable “fire-and-forget” background tasks, gather them in a collection:


This is not needed if you use the Home Assistant apis

bdraco commented on homeassistant/components/eq3btsmart/climate.py:
```diff
@@ -0,0 +1,391 @@
+        if self._cancel_timer:
+            self._cancel_timer()
+
+    async def _async_scan_loop(self, now: datetime | None = None) -> None:
+        """Scan for data."""
+
```
It would be better to setup the scan loop in `async_setup_entry` so it can be torn down there when the entry is unloaded.

You can use the dispatcher (async_dispatcher_send and async_dispatcher_connect) to get data into the entity.

bdraco reviewed (changes requested):
Please see comments above

EuleMitKeule commented:
@bdraco Thank you for all your work! I will fix all that in the next days. Sorry for making so many mistakes. 😅

EuleMitKeule committed:
Handle connection and update in setup entry, fix tests, simplify

EuleMitKeule commented:
@bdraco I fixed everything except the two review comments that are still unresolved because I think they are invalid. 
The update loop and reconnection logic is now handled in `__init__.py` using `entry.async_create_background_task` and `async_dispatcher_send`.
Would be great if you could take another look!

EuleMitKeule committed:
Merge branch 'dev' into eq3btsmart

bdraco commented on homeassistant/components/eq3btsmart/climate.py:
```diff
@@ -0,0 +1,306 @@
+
+class Eq3Climate(Eq3Entity, ClimateEntity):
+    """Climate entity to represent a eQ-3 thermostat."""
+
```
```suggestion

    _attr_has_entity_name = True
   _attr_name = None
   _attr_supported_features = (
            ClimateEntityFeature.TARGET_TEMPERATURE
            | ClimateEntityFeature.PRESET_MODE
            | ClimateEntityFeature.TURN_OFF
            | ClimateEntityFeature.TURN_ON
    )
    _attr_min_temp = EQ3BT_OFF_TEMP
    _attr_max_temp = EQ3BT_MAX_TEMP
```


Please put anything that never changes here and remove from `__init__`

`ruff` may complain about the formatting of this suggestion

EuleMitKeule replied:
Done.

EuleMitKeule committed:
Remove unused adapter enum

EuleMitKeule committed:
Remove constant attributes from init

bdraco commented on homeassistant/components/eq3btsmart/__init__.py:
```diff
@@ -0,0 +1,138 @@
+    """Reconnect the thermostat."""
+
+    eq3_config_entry: Eq3ConfigEntry = hass.data[DOMAIN][entry.entry_id]
+
```
https://github.com/home-assistant/core/pull/109291/files#r1480123746

bdraco commented on homeassistant/components/eq3btsmart/__init__.py:
```diff
@@ -0,0 +1,138 @@
+    entry.async_on_unload(entry.add_update_listener(update_listener))
+    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
+
+    entry.async_create_background_task(
+        hass, _run_thermostat(hass, entry), entry.entry_id
+    )
```
This needs to be cancelled when the config entry in unloaded

EuleMitKeule replied:
But the docs on `async_create_background_task` say "Background tasks are automatically canceled when config entry is unloaded."

bdraco replied:
Oh right. I forgot we added that.

All good than.

EuleMitKeule replied:
Do you mean like this?

```py
async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Handle config entry unload."""

    if unload_ok := await hass.config_entries.async_unload_platforms(entry, PLATFORMS):
        eq3_config_entry: Eq3ConfigEntryData = hass.data[DOMAIN].pop(entry.entry_id)

        eq3_config_entry.run_task.cancel()

        with suppress(asyncio.CancelledError):
            await eq3_config_entry.run_task
        # ...
    ```

bdraco commented on homeassistant/components/eq3btsmart/__init__.py:
```diff
@@ -0,0 +1,138 @@
+        ble_device=device,
+    )
+
+    eq3_config_entry = Eq3ConfigEntry(eq3_config=eq3_config, thermostat=thermostat)
```
```suggestion
    eq3_config_entry = Eq3ConfigEntry(eq3_config=eq3_config, thermostat=thermostat)
```

Let's call this `Eq3ConfigEntryData` so its clear what it is

bdraco commented on homeassistant/components/eq3btsmart/climate.py:
```diff
@@ -0,0 +1,304 @@
+
+        super().__init__(eq3_config, thermostat)
+
+        self._thermostat.register_update_callback(self._async_on_updated)
```
Please avoid side effects in `__init__`
Use the dispatcher instead.

bdraco commented on homeassistant/components/eq3btsmart/climate.py:
```diff
@@ -0,0 +1,304 @@
+    def _async_on_disconnected(self, mac_address: str) -> None:
+        if mac_address == self._eq3_config.mac_address:
+            self._attr_available = False
+            self.async_schedule_update_ha_state(force_refresh=True)
```
```suggestion
            self.async_write_ha_state()
```

bdraco commented on homeassistant/components/eq3btsmart/climate.py:
```diff
@@ -0,0 +1,304 @@
+    def _async_on_connected(self, mac_address: str) -> None:
+        if mac_address == self._eq3_config.mac_address:
+            self._attr_available = True
+            self.async_schedule_update_ha_state(force_refresh=True)
```
```suggestion
            self.async_write_ha_state()
```

bdraco commented on homeassistant/components/eq3btsmart/climate.py:
```diff
@@ -0,0 +1,304 @@
+        if self._thermostat.device_data is not None:
+            self._async_on_device_updated()
+
+        self.async_schedule_update_ha_state()
```
I think you can write state directly here as no update appears to be needed

`self.async_write_ha_state()`

bdraco commented on homeassistant/components/eq3btsmart/climate.py:
```diff
@@ -0,0 +1,304 @@
+    @callback
+    def _async_on_status_updated(self) -> None:
+        """Handle updated status from the thermostat."""
+
```
Since you have a single update point. I think it would make this easier to follow if you got rid of `target_temperature` and `current_temperature` and set them as `_attr_target_temperature` and `_attr_current_temperature` here (you may want to convert them to normal functions to keep the return pattern)

bdraco commented on homeassistant/components/eq3btsmart/climate.py:
```diff
@@ -0,0 +1,304 @@
+
+    async_add_entities(
+        [Eq3Climate(eq3_config_entry.eq3_config, eq3_config_entry.thermostat)],
+        update_before_add=False,
```
```suggestion
```

This is the default

bdraco commented on .coveragerc:
```diff
@@ -356,6 +356,11 @@ omit =
+    homeassistant/components/eq3btsmart/__init__.py
+    homeassistant/components/eq3btsmart/climate.py
+    homeassistant/components/eq3btsmart/const.py
+    homeassistant/components/eq3btsmart/eq3_entity.py
```
Lets call this `entity.py` since its already in `eq3btsmart`

EuleMitKeule committed:
Various fixes of reviewed minor problems

bdraco commented on homeassistant/components/eq3btsmart/__init__.py:
```diff
@@ -0,0 +1,146 @@
+    """Reconnect the thermostat."""
+
+    eq3_config_entry: Eq3ConfigEntryData = hass.data[DOMAIN][entry.entry_id]
+    thermostat: Thermostat = eq3_config_entry.thermostat
+    mac_address: str = eq3_config_entry.eq3_config.mac_address
+    scan_interval: int = eq3_config_entry.eq3_config.scan_interval
+    name: str = eq3_config_entry.eq3_config.name
```
```suggestion
    thermostat = eq3_config_entry.thermostat
    name = eq3_config_entry.eq3_config.name
    mac_address = eq3_config_entry.eq3_config.mac_address
    scan_interval = eq3_config_entry.eq3_config.scan_interval
```

The type hints are already defined in the model

bdraco commented on homeassistant/components/eq3btsmart/__init__.py:
```diff
@@ -0,0 +1,146 @@
+    """Run the thermostat."""
+
+    eq3_config_entry: Eq3ConfigEntryData = hass.data[DOMAIN][entry.entry_id]
+    thermostat: Thermostat = eq3_config_entry.thermostat
+    name: str = eq3_config_entry.eq3_config.name
+    mac_address: str = eq3_config_entry.eq3_config.mac_address
+    scan_interval: int = eq3_config_entry.eq3_config.scan_interval
```
```suggestion
    thermostat = eq3_config_entry.thermostat
    name = eq3_config_entry.eq3_config.name
    mac_address = eq3_config_entry.eq3_config.mac_address
    scan_interval = eq3_config_entry.eq3_config.scan_interval
```

EuleMitKeule committed:
Remove unneeded type hints

bdraco commented on homeassistant/components/eq3btsmart/__init__.py:
```diff
@@ -0,0 +1,146 @@
+            await asyncio.sleep(scan_interval)
+            continue
+
+        _LOGGER.info(
```
```suggestion
        _LOGGER.debug(
```

EuleMitKeule replied:
I thought an info log was required, since it says in https://developers.home-assistant.io/docs/integration_quality_scale_index/#silver-
> Log a warning once when unavailable, log once when reconnected.

But I guess I'll change it to debug.

bdraco commented on homeassistant/components/eq3btsmart/__init__.py:
```diff
@@ -0,0 +1,146 @@
+                await _reconnect_thermostat(hass, entry)
+                continue
+
+            _LOGGER.error(
+                "[%s] Error updating eQ-3 device: %s",
+                name,
+                e,
+            )
```
Ideally this only logs once when it fails and does not log again until its failed after it has succeeded so we don't flood the log when its offline

https://developers.home-assistant.io/docs/integration_quality_scale_index?_highlight=logs&_highlight=once#silver-

> Handles device/service unavailable. Log a warning once when unavailable, log once when reconnected.

EuleMitKeule replied:
I think this is already the case, since that message only gets logged when an error occurs while the thermostat is connected. If it disconnects, the if-statement above runs and skips the loop iteration after reconnecting.

bdraco commented on homeassistant/components/eq3btsmart/climate.py:
```diff
@@ -0,0 +1,320 @@
+    async def async_added_to_hass(self) -> None:
+        """Run when entity about to be added to hass."""
+
+        self._thermostat.register_update_callback(self._async_on_updated)
```
How does this get removed when the integration is unloaded/reloaded?

EuleMitKeule replied:
It does not, I thought the `Thermostat` object gets recreated when the integration is reloaded, but that's obviously not the case. I'll implement a `unregister_update_callback` in the library and run that in `async_will_remove_from_hass`.

bdraco replied:
I think `register_connection_callback` probably has the same issue?

EuleMitKeule replied:
It had, but it's not used anymore since I handle the connection and reconnection logic in the integration now.

bdraco commented on homeassistant/components/eq3btsmart/climate.py:
```diff
@@ -0,0 +1,320 @@
+        previous_mode = self.hvac_mode
+        self._attr_hvac_mode = hvac_mode
+
+        self.async_schedule_update_ha_state()
```
I think you can write state here directly as well

bdraco commented on homeassistant/components/eq3btsmart/climate.py:
```diff
@@ -0,0 +1,320 @@
+            _LOGGER.error("[%s] Failed setting HVAC mode", self._eq3_config.name)
+            self._attr_hvac_mode = previous_mode
+
+        self.async_schedule_update_ha_state()
```
I think you can write state here directly as well

bdraco commented on homeassistant/components/eq3btsmart/config_flow.py:
```diff
@@ -0,0 +1,131 @@
+
+        self.name = ""
+        self.mac = ""
+        self.rssi = 0
```
I'm not sure we should show the rssi here as it doesn't seem relevant unless its failing to setup

bdraco commented on homeassistant/components/eq3btsmart/entity.py:
```diff
@@ -0,0 +1,18 @@
+from .models import Eq3Config
+
+
+class Entity(HassEntity):
```
```suggestion
class Eq3BTEntity(Entity):
```

bdraco replied:
Or some other name..

bdraco commented on homeassistant/components/eq3btsmart/entity.py:
```diff
@@ -0,0 +1,18 @@
+
+from eq3btsmart.thermostat import Thermostat
+
+from homeassistant.helpers.entity import Entity as HassEntity
```
```suggestion
from homeassistant.helpers.entity import Entity
```

Importing it as a different name makes this much harder for maintainers for find these

bdraco commented on homeassistant/components/eq3btsmart/entity.py:
```diff
@@ -0,0 +1,18 @@
+
+        self._eq3_config = eq3_config
+        self._thermostat = thermostat
+        self._attr_name = self._eq3_config.name
```
Since we are using has_entity_name should we remove this?

EuleMitKeule replied:
This will be used by the other platforms that I'll add later, but I can remove it for now.

bdraco commented on homeassistant/components/eq3btsmart/climate.py:
```diff
@@ -0,0 +1,320 @@
+    """Climate entity to represent a eQ-3 thermostat."""
+
+    _attr_has_entity_name: bool = True
+    _attr_name: str | None = None
```
```suggestion
    _attr_name = None
```

typing is already set by the base class. 

Please remove redundant type hints

EuleMitKeule replied:
My IDE does not understand this, but I'll remove them.

![grafik](https://github.com/home-assistant/core/assets/47421406/156964c7-b358-4541-9837-1c413e998dc5)

EuleMitKeule committed:
Unregister update callback, remove rssi, minor fixes

bdraco commented on homeassistant/components/eq3btsmart/__init__.py:
```diff
@@ -0,0 +1,146 @@
+    await hass.config_entries.async_reload(entry.entry_id)
+
+
+async def _run_thermostat(hass: HomeAssistant, entry: ConfigEntry) -> None:
```
```suggestion
async def _async_run_thermostat(hass: HomeAssistant, entry: ConfigEntry) -> None:
```

bdraco commented on homeassistant/components/eq3btsmart/__init__.py:
```diff
@@ -0,0 +1,146 @@
+                    SIGNAL_THERMOSTAT_DISCONNECTED,
+                    mac_address,
+                )
+                await _reconnect_thermostat(hass, entry)
```
```suggestion
                await _async_reconnect_thermostat(hass, entry)
```

bdraco commented on homeassistant/components/eq3btsmart/__init__.py:
```diff
@@ -0,0 +1,146 @@
+    mac_address = eq3_config_entry.eq3_config.mac_address
+    scan_interval = eq3_config_entry.eq3_config.scan_interval
+
+    await _reconnect_thermostat(hass, entry)
```
```suggestion
    await _async_reconnect_thermostat(hass, entry)
```

bdraco commented on homeassistant/components/eq3btsmart/__init__.py:
```diff
@@ -0,0 +1,146 @@
+        await asyncio.sleep(scan_interval)
+
+
+async def _reconnect_thermostat(hass: HomeAssistant, entry: ConfigEntry) -> None:
```
```suggestion
async def _async_reconnect_thermostat(hass: HomeAssistant, entry: ConfigEntry) -> None:
```

bdraco commented on homeassistant/components/eq3btsmart/climate.py:
```diff
@@ -0,0 +1,313 @@
+        previous_temperature = self._target_temperature
+        self._target_temperature = temperature
+
+        self.async_schedule_update_ha_state()
```
Looks like more cases that can be replaced with `async_write_ha_state` in this function

bdraco commented on homeassistant/components/eq3btsmart/config_flow.py:
```diff
@@ -0,0 +1,127 @@
+        and all(int(part, 16) < 256 for part in mac.split(":") if part)
+    )
+
+
+def validate_name(name: str) -> bool:
+    """Return whether or not given value is a valid name."""
+
+    return bool(name)
```
This looks like a useless delegation

bdraco replied:
```suggestion
```

bdraco commented on homeassistant/components/eq3btsmart/config_flow.py:
```diff
@@ -0,0 +1,127 @@
+        name = user_input[CONF_NAME]
+        mac_address = user_input[CONF_MAC]
+
+        if not validate_name(name):
```
```suggestion
        if not name:
```

bdraco commented on homeassistant/components/eq3btsmart/config_flow.py:
```diff
@@ -0,0 +1,127 @@
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle a flow initialized by the user."""
+
```
If user flows we usually enumerate the devices in range and see if they are supported and than offer the user a device to pick.  Switchbot has a good example of this

bdraco replied:
Govee_ble does as well but it for passive devices only so it might not be as good a reference

EuleMitKeule replied:
Not sure this is viable in this case, because the thermostats disable their bluetooth radio frequently and most often are not advertising at all when the user creates the entry.

bdraco replied:
Please add a comment to that effect so it doesn't come up in future reviews as thats not a very common case

EuleMitKeule commented on homeassistant/components/eq3btsmart/config_flow.py:
```diff
@@ -0,0 +1,127 @@
+                errors=errors,
+            )
+
+        name = user_input[CONF_NAME]
```
I could set the default device name to be the mac address. I don't think this is very user friendly, but I understand your argument.

bdraco replied:
If the device is in range you could get the local name via https://developers.home-assistant.io/docs/core/bluetooth/api#fetching-the-latest-bluetoothserviceinfobleak-for-a-device and use that as the default, and fallback to the address if you don't have it

EuleMitKeule replied:
The local name is always just `CC-RT-BLE` so I'll rather use the mac since that is unique.

bdraco commented on homeassistant/components/eq3btsmart/const.py:
```diff
@@ -0,0 +1,78 @@
+DEFAULT_TARGET_TEMP_SELECTOR = TargetTemperatureSelector.TARGET
+DEFAULT_SCAN_INTERVAL = 10  # seconds
+
+SIGNAL_THERMOSTAT_DISCONNECTED = f"{DOMAIN}.thermostat_disconnected"
+SIGNAL_THERMOSTAT_CONNECTED = f"{DOMAIN}.thermostat_connected"
```
We need to include a unique identifier in these or it will dispatch to every instance of the integration

example https://github.com/home-assistant/core/blob/6519b24319c571580294ade48738123d35dc7b12/homeassistant/components/esphome/entry_data.py#L148

bdraco commented on homeassistant/components/eq3btsmart/climate.py:
```diff
@@ -0,0 +1,313 @@
+        """Set new target temperature."""
+
+        if ATTR_HVAC_MODE in kwargs:
+            mode = kwargs.get(ATTR_HVAC_MODE)
+
+            if mode is None:
```
```suggestion
            if (mode := kwargs.get(ATTR_HVAC_MODE)) is None:
```

bdraco commented on homeassistant/components/eq3btsmart/climate.py:
```diff
@@ -0,0 +1,313 @@
+
+            await self.async_set_hvac_mode(mode)
+
+        temperature = kwargs.get(ATTR_TEMPERATURE)
+
+        if temperature is None:
```
```suggestion
        if (temperature := kwargs.get(ATTR_TEMPERATURE)) is None:
```

bdraco commented on homeassistant/components/eq3btsmart/const.py:
```diff
@@ -0,0 +1,78 @@
+    LOW_BATTERY = "Low Battery"
+    WINDOW_OPEN = "Window"
+
+
+CONF_ADAPTER = "conf_adapter"
+CONF_CURRENT_TEMP_SELECTOR = "conf_current_temp_selector"
+CONF_TARGET_TEMP_SELECTOR = "conf_target_temp_selector"
+CONF_EXTERNAL_TEMP_SENSOR = "conf_external_temp_sensor"
```
These look unused at this point
```suggestion
```

EuleMitKeule committed:
Remove name and rssi from config flow, use unique signals

EuleMitKeule committed:
Slugify config entry title

EuleMitKeule commented:
I did everything as requested. But now for some reason when I change the device name the entity is not automatically renamed, even though the device, entity and config entry all have the exact same name. This worked before. Any ideas how I can fix this?

bdraco commented:
> I did everything as requested. But now for some reason when I change the device name the entity is not automatically renamed, even though the device, entity and config entry all have the exact same name. This worked before. Any ideas how I can fix this?

How are you changing the name?

joostlek commented on homeassistant/components/eq3btsmart/climate.py:
```diff
@@ -0,0 +1,316 @@
+        self._attr_preset_mode = None
+        self._attr_unique_id = format_mac(self._eq3_config.mac_address)
+        self._attr_device_info = DeviceInfo(
+            name=slugify(self._eq3_config.mac_address),
+            manufacturer=MANUFACTURER,
+            model=DEVICE_MODEL,
+            sw_version=None,
+            serial_number=None,
+            connections={(CONNECTION_BLUETOOTH, self._eq3_config.mac_address)},
```
```suggestion
            name=slugify(eq3_config.mac_address),
            manufacturer=MANUFACTURER,
            model=DEVICE_MODEL,
            connections={(CONNECTION_BLUETOOTH, eq3_config.mac_address)},
```

joostlek commented on homeassistant/components/eq3btsmart/climate.py:
```diff
@@ -0,0 +1,316 @@
+
+        super().__init__(eq3_config, thermostat)
+
+        self._target_temperature: float | None = None
+        self._attr_available = False
+        self._attr_hvac_mode = None
+        self._attr_hvac_action = None
+        self._attr_preset_mode = None
```
Can be set outside of the constructor

EuleMitKeule replied:
What's the convention on this? @bdraco said I should only move the attributes that never change outside the constructor.

joostlek commented on homeassistant/components/eq3btsmart/climate.py:
```diff
@@ -0,0 +1,316 @@
+        self._attr_hvac_mode = None
+        self._attr_hvac_action = None
+        self._attr_preset_mode = None
+        self._attr_unique_id = format_mac(self._eq3_config.mac_address)
```
```suggestion
        self._attr_unique_id = format_mac(eq3_config.mac_address)
```

joostlek commented on homeassistant/components/eq3btsmart/climate.py:
```diff
@@ -0,0 +1,316 @@
+class Eq3Climate(Eq3Entity, ClimateEntity):
+    """Climate entity to represent a eQ-3 thermostat."""
+
+    _attr_has_entity_name = True
```
Please move this one to the parent entity, as all child entities will have to use this

EuleMitKeule commented:
> How are you changing the name?

@bdraco Clicking the pen icon on the top right corner in the device UI.

EuleMitKeule committed:
Remove attributes from constructor

EuleMitKeule commented:
The type hints on `_attr_hvac_mode`, `_attr_preset_mode` and `_attr_hvac_action` are required for mypy not to fail. The type hints in `ClimateEntity` do not seem to apply, maybe because of multiple inheritance?

EuleMitKeule committed:
Slugify entry title in tests

bdraco commented:
> > How are you changing the name?
> 
> @bdraco Clicking the pen icon on the top right corner in the device UI.

It might be a frontend bug.  I had someone else tell me the same thing for a different integration last week.  I haven't found an issue report for it yet though

bdraco commented on homeassistant/components/eq3btsmart/config_flow.py:
```diff
@@ -0,0 +1,100 @@
+    ) -> FlowResult:
+        """Handle flow start."""
+
+        self._async_abort_entries_match({CONF_MAC: self.mac_address})
```
It looks like we don't need to store `CONF_MAC` since its the same as the unique id

bdraco reviewed (dismissed):
Looks pretty close to ready to merge.  One more comment left above

EuleMitKeule committed:
Remove obsolete check in config flow

autinerd commented on homeassistant/components/eq3btsmart/strings.json:
```diff
@@ -0,0 +1,19 @@
+{
+  "config": {
+    "abort": {
+      "already_configured": "Device is already configured"
```
```suggestion
      "already_configured": "[%key:common::config_flow::abort::already_configured_device%]"
```

autinerd reviewed (changes requested):
One small comment from me as well, but looks good :)

EuleMitKeule committed:
Use common string

Co-authored-by: Sid <27780930+autinerd@users.noreply.github.com>

bdraco commented:
It looks like this comment is still outstanding https://github.com/home-assistant/core/pull/109291#discussion_r1481583640

EuleMitKeule committed:
Do not store mac in config entry

EuleMitKeule commented:
@bdraco Should be fixed now!

bdraco commented on tests/components/eq3btsmart/test_config_flow.py:
```diff
@@ -0,0 +1,132 @@
+
+    assert result["type"] == FlowResultType.CREATE_ENTRY
+    assert result["title"] == slugify(MAC)
+    assert result["data"] == {}
```
```suggestion
    assert result["data"] == {}
    assert result["context"]["unique_id"] == MAC    
```

It would be good to assert the unique id is set as well

This suggestion may need to be adjusted

EuleMitKeule committed:
Use assert for type checking

EuleMitKeule committed:
Assert unique id is set

bdraco committed:
Merge branch 'dev' into eq3btsmart

bdraco reviewed (approved):
LGTM.

I'm pretty low on sleep so I may have missed something so I'll leave this until someone else can take a look or I've had enough sleep that I feel comfortable merging

autinerd reviewed (approved):
Looks good to me as well.

bdraco committed:
Merge branch 'dev' into eq3btsmart

EuleMitKeule commented:
@bdraco Do you think there is any chance this could be included in 2024.3 or is it too late now?

EuleMitKeule commented:
I was really hoping for this one to be included in 2024.4 since there has been approval and everything since early February. A little demotivating, but I guess this process can take some time.

Also there are more features in the pipeline like options flow and additional platforms that expose more configuration and control entities.

bdraco committed:
Merge branch 'dev' into eq3btsmart

bdraco committed:
lint

bdraco committed:
lint

bdraco committed:
cleanups

bdraco commented on homeassistant/components/eq3btsmart/climate.py:
```diff
@@ -0,0 +1,312 @@
+
+                return float(self._thermostat.status.target_temperature.value)
+
+    def _get_current_preset_mode(self) -> str:
+        """Return the current preset mode."""
+
+        if self._thermostat.status is None:
+            return PRESET_NONE
+        if self._thermostat.status.is_window_open:
+            return Preset.WINDOW_OPEN
+        if self._thermostat.status.is_boost:
+            return Preset.BOOST
+        if self._thermostat.status.is_low_battery:
+            return Preset.LOW_BATTERY
+        if self._thermostat.status.is_away:
+            return Preset.AWAY
+        if self._thermostat.status.operation_mode is OperationMode.ON:
+            return Preset.OPEN
+        if self._thermostat.status.presets is None:
+            return PRESET_NONE
+        if (
+            self._thermostat.status.target_temperature
+            == self._thermostat.status.presets.eco_temperature
+        ):
+            return Preset.ECO
+        if (
+            self._thermostat.status.target_temperature
+            == self._thermostat.status.presets.comfort_temperature
+        ):
+            return Preset.COMFORT
+
+        return PRESET_NONE
```
```suggestion
    def _get_current_preset_mode(self) -> str:
        """Return the current preset mode."""

        if (status := self._thermostat.status) is None:
            return PRESET_NONE
        if status.is_window_open:
            return Preset.WINDOW_OPEN
        if status.is_boost:
            return Preset.BOOST
        if status.is_low_battery:
            return Preset.LOW_BATTERY
        if status.is_away:
            return Preset.AWAY
        if status.operation_mode is OperationMode.ON:
            return Preset.OPEN
        if status.presets is None:
            return PRESET_NONE
        if status.target_temperature == status.presets.eco_temperature:
            return Preset.ECO
        if status.target_temperature == status.presets.comfort_temperature:
            return Preset.COMFORT

        return PRESET_NONE
```

bdraco committed:
bring up to date

bdraco commented on tests/components/eq3btsmart/test_config_flow.py:
```diff
@@ -0,0 +1,135 @@
+        )
+        await hass.async_block_till_done()
+
+    assert result["type"] == FlowResultType.ABORT
```
```suggestion
    assert result["type"] is FlowResultType.ABORT
```

bdraco commented on tests/components/eq3btsmart/test_config_flow.py:
```diff
@@ -0,0 +1,135 @@
+        )
+        await hass.async_block_till_done()
+
+    assert result["type"] == FlowResultType.CREATE_ENTRY
```
```suggestion
    assert result["type"] is FlowResultType.CREATE_ENTRY
```

bdraco commented on tests/components/eq3btsmart/test_config_flow.py:
```diff
@@ -0,0 +1,135 @@
+        )
+        await hass.async_block_till_done()
+
+        assert result["type"] == FlowResultType.CREATE_ENTRY
```
```suggestion
        assert result["type"] is FlowResultType.CREATE_ENTRY
```

bdraco commented on tests/components/eq3btsmart/test_config_flow.py:
```diff
@@ -0,0 +1,135 @@
+        )
+        await hass.async_block_till_done()
+
+        assert result["type"] == FlowResultType.FORM
```
```suggestion
        assert result["type"] is FlowResultType.FORM
```

bdraco commented on tests/components/eq3btsmart/test_config_flow.py:
```diff
@@ -0,0 +1,135 @@
+        )
+        await hass.async_block_till_done()
+
+    assert result["type"] == FlowResultType.CREATE_ENTRY
```
```suggestion
    assert result["type"] is FlowResultType.CREATE_ENTRY
```

bdraco committed:
Apply suggestions from code review

bdraco commented:
Sorry I didn't have time to look at this one sooner.  beta has been a bit smoother than expected so I had some free cycles today to give this another quick look.

bdraco commented:
Thanks @EuleMitKeule

EuleMitKeule commented:
@bdraco Thanks so much for all the work in reviewing and everything!
